Sign up screen

FadhlanR opened the pull request:
Ticket: CS-6138



https://github.com/cardstack/boxel/assets/12637010/7c1e1b35-65af-4ede-a62f-5eaf3d698c01

FadhlanR committed:
Sign up screen

FadhlanR committed:
Make auth overflow auto

FadhlanR committed:
fix autofill issue

FadhlanR committed:
Update email template

FadhlanR committed:
fix tests

FadhlanR committed:
fix matrix test

jurgenwerk commented on packages/boxel-ui/addon/src/components/input-group/accessories/index.gts:
```diff
@@ -0,0 +1,123 @@
+  Element: HTMLButtonElement | HTMLAnchorElement;
+}
+
+export const Button: TemplateOnlyComponent<ButtonSignature> = <template>
```
Was the input group copied from the cardstack monorepo? Any adjustments made?

FadhlanR replied:
yes I copied this from cardstack monorepo. The style is different.

<img width="516" alt="Screen Shot 2023-12-13 at 11 21 33" src="https://github.com/cardstack/boxel/assets/12637010/918d7730-54c9-45aa-b253-997440862b5d">
<img width="1154" alt="Screen Shot 2023-12-13 at 11 21 26" src="https://github.com/cardstack/boxel/assets/12637010/82171824-93f6-4b9f-b537-270e921346fb">

jurgenwerk commented on packages/matrix/docker/synapse/dev/homeserver.yaml:
```diff
@@ -87,6 +87,9 @@ email:
   notif_for_new_users: false
   client_base_url: "http://localhost:4200"
-  validation_token_lifetime: "15m"
+  validation_token_lifetime: "48h"
```
Why change to 48h?

habdelra replied:
the ticket specifically asked for 48 hours. albeit it's kinda weird since we also told the user not to close their browser while we send the validation email. seems like that should be an hour at most...

jurgenwerk replied:
Agree, it seems smaller time units are suitable for this mechanism.

jurgenwerk commented on packages/matrix/docker/synapse/templates/registration.html:
```diff
@@ -0,0 +1,89 @@
+  </div>
+</div>
+</body>
+</html>
```
Looks messy, can you run this through a html formatter?

jurgenwerk commented on packages/matrix/docker/synapse/templates/registration.html:
```diff
@@ -0,0 +1,89 @@
+  </div>
+  <div class="content">
+    <span class="title">Verification your email to register your Boxel account</span>
+    <p>Please confirm that <b>{{ email }}</b> is your e-mail address by clicking on the button below or use this link <a href="{{ link }}">{{ link }}</a> within 48 hours.</p>
```
The link is too long to show it like that in a sentence, consider using `<a href="{{ link }}">this link</a>`

<img width="1057" alt="image" src="https://github.com/cardstack/boxel/assets/273660/30316273-9614-4b59-b9cf-0a3e9014663f">

jurgenwerk replied:
"Please confirm that is your email address" -> "Please confirm that this is your email address"

jurgenwerk commented on packages/matrix/docker/synapse/templates/registration.html:
```diff
@@ -0,0 +1,89 @@
+    <span class="header__title">Boxel</span>
+  </div>
+  <div class="content">
+    <span class="title">Verification your email to register your Boxel account</span>
```
Verification -> Verify

habdelra commented on packages/host/app/components/matrix/register-user.gts:
```diff
@@ -295,62 +369,56 @@ export default class RegisterUser extends Component<Signature> {
+      this.nameError ||
+      this.usernameError ||
+      this.passwordError ||
+      this.confirmPasswordError
```
this is hard to reason about. i think what you mean to say is that if there is a missing field or there is an error then the register button is disabled. can we break this down into a getter `get hasMissingField()` and `get hasError()`, and then just make this `return this.hasMissingField || this.hasError` so that it is more clear what the intent is here?

habdelra commented on packages/host/app/components/matrix/register-user.gts:
```diff
@@ -386,35 +464,78 @@ export default class RegisterUser extends Component<Signature> {
+    if (!this.email) {
+      this.emailError = 'Email address is missing';
+    } else if (!/^[\w-\.]+@([\w-]+\.)+[\w-]{2,4}$/.test(this.email)) {
+      this.emailError = 'Email address is not a valid email';
```
this is not the hill to die on :-) many people have spend many thousands of man years trying to validate email from inspection. lets not fight this battle. the ultimately the way to really know if an email is valid is to just try sending the email. let's just check if its missing and let that be it.

habdelra commented:
one nice suggestion would be to write a script that makes a test registration token, just like we make a test user--i think the other devs will appreciate that. There are examples of how to create a registration token in our matrix tests

habdelra commented on packages/host/app/components/matrix/register-user.gts:
```diff
@@ -16,241 +15,313 @@ import { v4 as uuidv4 } from 'uuid';
+            to complete registration.</span>
+          <ul class='email-validation-instruction'>
+            <li>Leave this window while we verify your email</li>
+            <li>This scree will update once your email is verified</li>
```
typo: `scree`

habdelra commented on packages/host/app/components/matrix/register-user.gts:
```diff
@@ -16,241 +15,313 @@ import { v4 as uuidv4 } from 'uuid';
+          <span class='title' data-test-email-validation>Please check your email
+            to complete registration.</span>
+          <ul class='email-validation-instruction'>
+            <li>Leave this window while we verify your email</li>
```
Is this supposed to say "Leave this window _open_ while we verify your email"

habdelra commented:
A useful instruction for any developer looking at this is that you need to make sure to stop your synapse first and then run `start:all`--otherwise we won't pickup the new matrix config. the docker container that synapse runs in is only stopped from 
```
packages/matrix $ pnpm stop:synapse
```

habdelra commented:
The email validation screen is not branded, is that intentional? may be worth a followup PR:
![image](https://github.com/cardstack/boxel/assets/61075/2aaccc19-66d3-4b06-849f-44a242e348ec)

habdelra commented:
I don't think we are capturing the display name correctly. The name that I enter in the registration form is not the same name that I see when l open that chat side panel. Can you look into this?

habdelra commented:
When you resend a validation email we cycle thru the ask registration token state. please fix this

![reg-flash](https://github.com/cardstack/boxel/assets/61075/2575f65a-4261-4751-8530-79aad81e485f)

FadhlanR committed:
Update based on feedback

FadhlanR committed:
Add script to register test token

FadhlanR commented:
@jurgenwerk @habdelra I've made some updates:

1. I've addressed formatting issues and corrected some typos in the `registration.html` file.
2. Added `registration_success.html` and `registration_failure.html` to improve the email validation screen.
3. Updated the `validation_token_lifetime` to be an hour.
4. Fixed `resend email` button and corrected some typos on that screen.
5. Fixed the display name issue.
6. Introduced a script for registering a test token.


https://github.com/cardstack/boxel/assets/12637010/b9627154-d8a5-49fc-a828-30b179fbff55

FadhlanR committed:
update verfication email screenshot

habdelra commented on packages/matrix/tests/registration-with-token.spec.ts:
```diff
@@ -111,52 +114,46 @@ test.describe('User Registration w/ Token', () => {
     await assertLoggedIn(page, {
       userId: '@user2:localhost',
-      displayName: 'user2',
+      displayName: 'Test User',
```
thanks!

FadhlanR committed:
Merge branch 'main' into cs-6138-sign-up-screen

jurgenwerk commented on packages/matrix/docker/synapse/templates/registration.txt:
```diff
@@ -0,0 +1,5 @@
+Please confirm that {{ email }} is your e-mail address by clicking this link within 48 hours.
```
1 hour

jurgenwerk replied:
Or just leave out the time component. If the link becomes invalid the user will be informed about that right?

FadhlanR replied:
There's no time config we could read in the email registration template. I think I'll leave it out. User will be redirected to registration failed page.

jurgenwerk reviewed (approved):
Looks good but I suggest to either leave out time component hardcoded in emails, or read it from the config.

FadhlanR committed:
Remove time component
